game window now resizes based on sidebar size. no more manual offsets.

esotericist opened the pull request:
#### Summary
SUMMARY: Interface "game window now resizes based on sidebar size. no more manual offsets."

#### Purpose of change
There's been a steady-state of various 'slightly off, but only sometimes' issue reports since the UI rewrite, due to the change in having the main game window take up part of the screen (where the sidebar wasn't), to it taking up all of the screen (with the sidebar rendered on top).

In addition, while this doesn't show up in the Tiles build the same way due to how buffering works, in curses there was a "z-fighting" type effect where the play area and sidebar would be drawing over each other, creating a flickering problem.

Fixes #29619 
Fixes #28863 

I have suspicions about #29627 that this PR will probably address, although without a currently-available failstate example I can't prove it.
With #29593 , this doesn't... fix it, exactly? I think there's an error in the related logic that causes it to not actually flash the appropriate tile. But at least it won't be flashing somewhere off to the side anymore. See 'Additional Context' below.
If #29184 isn't fixed already, it should be by this point (but obviously confirmation will be needed).

#### Describe the solution
Adjusted `game::init_ui()` to adjust the main game window size based off of edge offsets, which are currently controlled by the panels code (and currently supports things taking up both the right and left edges of the screen).  Also added a call to `game::init_ui()` from the panels configuration screen (for when the layout changes)

Removed every existing reference to `sidebar_offset` because it's no longer the same concept as it used to be.

Further note on the `sidebar_offsets` struct in use here: The current game logic should work absolutely correctly with panels of arbitrary widths rendered on both the left and right sides of the screen, if the panels code (or some other UI mechanism) correctly sets the appropriate edge offsets. There shouldn't be any need to chase this down in game logic in the future, and it would be trivial to extend this to support top and bottom (such as for things like the "hotkey bar" in android, or if someone wanted UI panels that pinned to top/bottom of the screen) if needed.

#### Describe alternatives you've considered
We could have continued down the path of chasing down all of the offsets by hand, but that was both unsustainable and was overall increasing maintenance burden.

#### Additional context
I have tested this in the tiles build both with and without tiles enabled, as well as in the curses build. As far as I can tell, everything we've been struggling with for offsets recently should be properly addressed in a final, generalized fashion.

I tested: looking around, firing weapons at monsters, throwing weapons at monsters, making zones, driving vehicles. I'll be genuinely surprised at this point if there's anything still problematic.

A gif of me shooting a zombie hulk in true curses mode:
![Screen Recording 2019-04-16 at 7 17 55 PM](https://user-images.githubusercontent.com/1569754/56257317-b0cfce80-6080-11e9-8f29-81a7f317291b.gif)

It's a close up, but I can promise nothing else is showing up on-screen outside of that window.

codemime commented on src/game.cpp:
```diff
@@ -459,7 +459,7 @@ void game::init_ui( const bool resized )
     to_map_font_dimension( TERRAIN_WINDOW_WIDTH, TERRAIN_WINDOW_HEIGHT );
 
-    VIEW_OFFSET_X = 0;
+    VIEW_OFFSET_X = 0 + sidebar_offsets.left;
```
```suggestion
    VIEW_OFFSET_X = sidebar_offsets.left;
```

codemime commented on src/game.cpp:
```diff
@@ -6753,7 +6743,7 @@ void game::zones_manager()
             wrefresh( w_zones_info );
 
-            center = center - sidebar_offset;
+            center = center;
```
```suggestion          
```

codemime commented on src/game.cpp:
```diff
@@ -7134,14 +7124,14 @@ look_around_result game::look_around( catacurses::window w_info, tripoint &cente
     temp_exit_fullscreen();
 
-    center = center + sidebar_offset;
+    center = center;
```
```suggestion
```

codemime commented on src/handle_action.cpp:
```diff
@@ -132,8 +131,8 @@ input_context game::get_player_input( std::string &action )
-        int offset_x = ( u.posx() + u.view_offset.x ) - ( getmaxx( w_terrain ) / 2 ) + g->sidebar_offset.x;
-        int offset_y = ( u.posy() + u.view_offset.y ) - ( getmaxy( w_terrain ) / 2 ) + g->sidebar_offset.y;
+        int offset_x = ( u.posx() + u.view_offset.x ) - ( getmaxx( w_terrain ) / 2 );
+        int offset_y = ( u.posy() + u.view_offset.y ) - ( getmaxy( w_terrain ) / 2 );
```
Since you've touched these lines: there's no need for parentheses here </hint>.

esotericist replied:
I'm used to having parentheses for clarity for mathy things, but I can see how they're not strictly necessary. I can remove that.

codemime commented on src/panels.cpp:
```diff
@@ -1829,6 +1830,16 @@ const std::string panel_manager::get_current_layout_id() const
 void panel_manager::init()
 {
     load();
+    update_offsets( panel_manager::get_manager().get_current_layout().begin()->get_width() );
```
This is happening inside of one of `panel_manager`'s functions already, so there's no need to request the singleton here. Also, `get_current_layout()` might [theoretically] provide no panels whatsoever: dereferencing `begin()` would yield UB in this case.

esotericist replied:
Honestly I am terrible at comprehending OO stuff and I couldn't figure out how to make this work any other way.

ifreund replied:
```suggestion
    update_offsets( get_current_layout().begin()->get_width() );
```

esotericist replied:
I was under the impression this code path guaranteed a current layout was set

ifreund replied:
Yeah it I looked more closely and that's definitely the case

esotericist commented on src/panels.cpp:
```diff
@@ -1829,6 +1830,16 @@ const std::string panel_manager::get_current_layout_id() const
+    if( get_option<std::string>( "SIDEBAR_POSITION" ) == "left" ) {
+        g->sidebar_offsets.left = x;
+    } else {
+        g->sidebar_offsets.right = x;
```
I don't know what 'feature envy' means. 

as for the bidirectional dependency, i was keeping data approximately where it was before, I genuinely didn't know what else to do with it that'd be more correct, especially with how panels is structured

kevingranade replied:
https://softwareengineering.stackexchange.com/questions/212128/what-is-a-feature-envy-code-and-why-is-it-considered-a-code-smell

Basically, if the game owns these offsets, then panel manager should just inform game about the update and have it handle fetching the options etc.

Possibly just by having game::init_ui() accept a with argument, or by having game::uinit_ui() fetch it directly with panel_manager::get_manager().get_current_layout().begin()->get_width();

esotericist replied:
Moved the data into `panel_manager` as private members, added public getters. addresses the bidirectional dependency, improves clarity as to what the data is.

esotericist commented:
I'm not in a position to resolve the bidirectional dependency aspect right now, so I think this is as done as I can get it for the time being.

esotericist commented:
This is *definitely* not ready because I have a NPE case I'm trying to chase down.

esotericist commented:
So, I don't know if this is still "correct" (and in particular I'm aware that the panels do not *yet* provide both left and right side widths simultaneously), but `init_ui()` is now pretty uncaring about sidebar position and width which I think is an improvement. 

I also renamed the members in `panels.cpp` when I moved that data out of `game.cpp`, to be a bit clearer (there was a bit of a nomenclature confusion in discord discussion involving "offset" that was linked to how this stuff used to work). I'm hoping `width_left` and `width_right` are fairly clear in purpose within the context of the panels.

Of particular note is the incredibly massive mistake I made directly altering `VIEW_OFFSET_X` with the left sidebar width, because that gets translated to a different coordinate system directly thereafter. I now alter the call to `catacurses::newwin()` and this prevents an overflow later on when calling `invalidate_framebuffer()` in `sdltiles.cpp`. It also reflects how `init_ui()` handled this before the UI rework.

I'm honestly astonished I *ever* got it to work with the sidebar in the left in my previous tests. I am certain I tested it, I am not certain how it didn't explode violently the way it has for the last few hours (before I finally figured out my mistake on `VIEW_OFFSET_X`). Such stacktrace. Very confuse.

*Hopefully* there aren't any strong opinions on the state of this PR at this point. This is easily the most exhausting thing I've attempted for this game so far, but I hope it's still worth the effort.

ifreund commented on src/panels.cpp:
```diff
@@ -1808,6 +1808,7 @@ panel_manager::panel_manager()
 {
     current_layout_id = "labels";
     layouts = initialize_default_panel_layouts();
+
```
```suggestion
```

ifreund commented on src/panels.cpp:
```diff
@@ -1826,9 +1827,32 @@ const std::string panel_manager::get_current_layout_id() const
+        return width_right;
+    }
+    return width_left;
+}
```
Looks like `get_width_right()` and `get_width_left()` are exactly the same, which probably means that you don't need both.

esotericist replied:
They are not. They're reversed in which they provide, based on the state of sidebar.

ifreund replied:
Ok, I really can't read

ifreund commented on src/panels.cpp:
```diff
@@ -1826,9 +1827,32 @@ const std::string panel_manager::get_current_layout_id() const
+void panel_manager::update_offsets( int x )
+{
+    width_right = x;
+    width_left = 0;
```
looks like width left is always 0? If it works like this, why are we even storing this? Or is this a typo?

esotericist replied:
That's to allow for panels to decide to use some space on the right *and* space on the left if need be. It isn't currently used but my understanding is panels will soon have the ability to not be restricted to one side of the screen.

That's why there's two values, and why `get_width_right()` and `get_width_left()` are mirrors at present. It's to ensure that `game.cpp` doesn't have to know where the sidebar is, only which parts of the screen are available for its use after the reduction of space for the sidebar(s).

ifreund replied:
Cool, that makes sense to me

esotericist commented:
Aight. Whether this is usable as it is or not, I think I have to be done with this one. It's as ready as I'm able to make it.

ifreund commented:
Fixes #29741
